### DISCUSSION TITLE
Idea for the Group By example in the Introductory Tutorial -- The Boltzmann Wealth Model

### DISCUSSION THREAD

**Discussion by st346** [mesa/mesa#2576] [Ideas]
The [Group By example in the Introductory Tutorial -- The Boltzmann Wealth Model](https://mesa.readthedocs.io/stable/tutorials/intro_tutorial.html#group-by) sets up 3 ethnicities in the model, but then chooses just one at random and sets all agents to this ethnicity

        # Create a list of our different ethnicities
        ethnicities = ["Green", "Blue", "Mixed"]

        # Create agents
        MoneyAgent.create_agents(model=self, n=n, ethnicity=self.random.choice(ethnicities))

This results in a monotone chart. 
![image](https://github.com/user-attachments/assets/05a8dd85-9968-4345-a4d7-f79e561b5e66)

***On a separate issue I changed 'ethnicity' to 'allegiance', as I feel it is less contentious.***

I modified my example so that agents are created with no allegiance.

        # Create agents
        MoneyAgent.create_agents(model=self, n=n)

This is reflected in the Agent Class **\_\_init\_\_**.  This class also has a new **choose_allegiance** function defined.

    def __init__(self, model, allegiance=None):
        super().__init__(model)
        self.wealth = 1
        self.allegiance = allegiance

    def choose_allegiance(self):
         # Create a list of our different allegiances
        allegiances = ["Green", "Blue", "Neutral"]
        self.allegiance=self.random.choice(allegiances)

After creating the Agents the model than asks them to choose their allegiance.

        # Create agents
        MoneyAgent.create_agents(model=self, n=n)
        
        # Get agents to choose allegiance
        self.agents.do("choose_allegiance")

When plotting the histogram I added the **multiple='stack'** parameter to produce a stacked histogram.

      # assign histogram colors
      palette = {"Green": "green", "Blue": "blue", "Mixed": "yellow"}
      sns.histplot(data=data, x="Wealth", multiple='stack', hue="Allegiance", discrete=True, palette=palette)
      g.set(title="Wealth distribution", xlabel="Wealth", ylabel="number of agents");

My step definition of the **step** in the model hasn't changed, apart from using allegiance and changing 'Mixed' to 'Neutral' in the if... else...

    def step(self):
        self.datacollector.collect(self)
        # groupby returns a dictionary of the different allegiances with a list of agents
        grouped_agents = model.agents.groupby("allegiance")

        for ally, similars in grouped_agents:
            if ally != "Neutral":
                similars.shuffle_do("give_money", similars)
            else:
                similars.shuffle_do("give_money", self.agents)  # This allows mixed to trade with anyone

I feel that the result is a more interesting histogram that shows the Neutrals end up poorest with 0 or 1, having given their money away, but not received any money from green or blue. 

![image](https://github.com/user-attachments/assets/e4d87758-4210-452c-a111-a1764cc8d706)

I feel this may be closer to what was originally intended in the **Group by** example, as the definition of the model's step can deal with a mix of ethnicities, but this doesn't actually occour in the original example.

**Comment by EwoutH** [comment]
@quaquel what do you think?

**Reply by quaquel** [reply]
I like the basic idea, but would simplify it a bit. For example, the `choose_allegiance` method is only called once if I understand it correctly, so drop this and just move the random selection into the `__init__` of the agent. Likewise, I would change the for loop naming a little bit to `for allegiance, group in grouped_agents`.
